### DISCUSSION TITLE
Load a json-ld string into a Grax schema

### DISCUSSION THREAD

**Discussion by ghost** [rdf-elixir/rdf-ex#25] [Q&A] [answered]
I am sorry to be so annoying, but I am learning the Semantic Web in the spare time and it is incredibly hard to find informations related to Elixir for such a niche topic.

Feel free to ignore this message or answer with a RTFM, really!

I have a very precise goal in mind: load data from a json-ld string into a Grax Schema.

Here is the code I am using, with comments:

```
# The loading of the vocab file works correctly
defmodule Cjp.Vocabs do
  use RDF.Vocabulary.Namespace

  defvocab SchemaOrg,
    base_iri: "http://schema.org/",
    file: "schemaorg-current-http.nt",
    ignore: [:"3DModel"]
end

# This is the schema definition, it will be a http://schema.org/Organization
defmodule Cjp.Schema.Organization do
  use Grax.Schema

  alias Cjp.Vocabs.SchemaOrg

  schema SchemaOrg.Organization do
    property :name, SchemaOrg.name
    property :description, SchemaOrg.description
    property :url, SchemaOrg.url
    property :logo, SchemaOrg.logo
  end
end

# This is the json-ld string I have as input, and I have full control over it:

input_str = """
{
  "@context": "http://schema.org/",
  "@type": "Organization",
  "name": "acme inc.",
  "description": "an evil company",
  "url": "https://www.acme.com",
  "logo": "https://www.acme.com/logo.png"
}
"""
```

It's not clear to me what are the steps I have to perform to achieve my goal.

Thanks for the patience, in case you reached this point :)

Cheers

**Comment by marcelotto** [comment]
No worries at all about asking questions! Learning Semantic Web technologies can indeed be challenging, especially when working with less common tech stacks like Elixir.

```elixir
# Your JSON-LD should include an @id like this:
input_str = """
{
  "@context": "http://schema.org/",
  "@type": "Organization",
  "@id": "https://www.acme.com#org",
  "name": "acme inc.",
  "description": "an evil company",
  "url": "https://www.acme.com",
  "logo": "https://www.acme.com/logo.png"
}
"""

# Then you can load it like this:
input_str
|> JSON.LD.read_string!()
|> Cjp.Schema.Organization.load(~I<https://www.acme.com#org>)
```

**Reply by ghost** [reply]
Thanks for the patience,
I changed the code a bit, to look like

```
input_str
|> JSON.LD.read_string!()
|> then(&RDF.Dataset.default_graph(&1))
|> Cjp.Schema.Organization.load(~I<https://www.acme.com#org>)
```

As the 1st argument of `load` should be a graph?

**Comment by marcelotto** [comment, marked as answer]
Oh, I thought `load/2` would accept datasets. Maybe I'll change that.

The `then`, however, shouldn't be necessary. 

BTW: if you're input indeed consists of multiple graphs and you want to load from all of them, you can aggregate the dataset with `RDF.Graph.new/1`:

```
input_str
|> JSON.LD.read_string!()
|> RDF.Graph.new()
|> Cjp.Schema.Organization.load(~I<https://www.acme.com#org>)
```
